clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -120,12 +116,18 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener liste
         changeSubscription(topics);
     }
 
-    public void changeSubscription(Collection<String> topicsToSubscribe) {
-        if (!this.subscription.equals(new HashSet<>(topicsToSubscribe))) {
-            this.subscription.clear();
-            this.subscription.addAll(topicsToSubscribe);
+    public void subscribeFromPattern(Set<String> topics) {
+        if (subscriptionType != SubscriptionType.AUTO_PATTERN)
+            throw new IllegalArgumentException("Attempt to subscribe from pattern while subscription type set to " +
+                    subscriptionType);
```

Comment:
@vahidhashemian It could use the same message I guess. This is more of a state assertion which we don't expect to reach the user because we control when it's called. I would have used an `assert` if that was more commonly used in the codebase.